### DISCUSSION TITLE
Custom Go func plugins to provide go-task plugin capability or easy extensionsibility

### DISCUSSION THREAD

**Discussion by sheldonhull** [go-task/task#486] [General]
I really like Go task for quickly setting up some projects with a simple execution of go, docker commands and more. 
Once I get beyond basics though it's hard to use in comparison to what I've used in the past.

I've got a background with PowerShell, which in this context is very robust and allows custom functions and very rich object interactions and pipelines. I know that PowerShell 7 (cross platform) doesn't fit in the model for this app, as it uses the `sh` package to do cross platform execution. However, it's bit difficult for me to start adding helpers like custom host colored output, argument builder for more complex app, etc. 

I also have used Mage, goyek, and others. 

I'm wondering if there could be a compromise in the simplicity of a yaml file for a task while still supporting some custom functions that could be very simple Go based functions as a "plugin". For example, I could include a `go-task.go` file that would use Pterm to provide nice console output or progress bars and then this plugin could be used by `go-task` like:

```yaml
- cmd: -|
    g.WriteConsole("My Nice Message using plugin")
   ls -ll
  echo foobar
```

It's a thought. If I could extend my tasks to do specific things without effort then I could simplify my task writing, and bind it together with cmd itself.

**Comment by sheldonhull** [comment]
@andreynering any tips? If it's possible already using the library directly can I use this in a similar fashion to Mage? I love the quick iteration with Task but want to write some reusable functions that handle logging say with Pterm, logic checks and other Go helper functions, while still leveraging the simplicity of Task if I can.

I've been using a combination of Goyek similar to Mage) and Task and am interested in capability to build more resumable Go helpers for Task

**Reply by andreynering** [reply]
Hi @sheldonhull,

Sorry for not responding... there has been many issues, and I have less time than I liked to answers all of them.

There's an old issue about using as a library open at https://github.com/go-task/task/issues/121.

You'll won't be able to add new template functions currently because Task does not expose any public method that would allow you to do that, but that's something we could consider doing...

Keep in mind that using as a library is not stable yet. This means that, although possible, you won't have the same backward compatibility guarantees you have when using the compiled binary. So I advice you to lock a specific version, and you may need to fix something here and there when upgrading.
